### PR TITLE
inference: Improve path matching performance

### LINKED ISSUE #48376: Improve performance of recognizer path filtering
On Sourcegraph.com, our largest auto indexing platform where we want to recheck about 3M repos in 12h, we are currently not hitting these numbers because _something_ in the recognizer chain is too slow to actually do 3M ops in 12h.

While not certain it will improve this, we definitely see in profiles that the majority of all CPU time spent by the worker is spent in filtering recognizer paths:

Image

Upon inspection, I noticed:

- gitserverClient.ListFiles does filtering client-side, so it makes sense that this overhead counts towards the worker CPU
- We use regexp matching on those file paths (which might make sense for some use-cases)
- From quick inspection of our lua recognizers however, our filters are extremely simple:

we usually do "path segment matching" or filename / file ending matching.

These can trivially be converted in much less expensive operations vs regexp matching, by doing simpler equivalent operations:

```lua
// for rust
  patterns = {
    pattern.new_path_basename "Cargo.toml",
  },
// for go
local exclude_paths = pattern.new_path_combine(shared.exclude_paths, {
  pattern.new_path_segment "vendor",
})
  patterns = {
    pattern.new_path_basename "go.mod",
    pattern.new_path_exclude(exclude_paths),
  },
  patterns = {
    pattern.new_path_extension "go",
    pattern.new_path_exclude(exclude_paths),
  },
// for python
  patterns = {
    pattern.new_path_basename "PKG-INFO",
  },
// and so on
```

```go
// pattern.new_path_basename
strings.HasSuffix("/" + basename)
// pattern.new_path_segment and pattern.new_path_exclude
contains(strings.Split(path, "/"), segment)
// 
```

you get the idea. If there are globs involved, I recently added a globbing library vs regexp here: https://github.com/sourcegraph/sourcegraph/pull/48314 with some benchmarks, too.

### PR DESCRIPTION
**Pull request by efritz**
Fixes #48376, but there may still be additional perf improvements after re-profiling.

This PR makes the following changes:

- Lua inference path patterns now construct sets of globs instead of regular expressions joined by an `|`.
- These globs are passed to `git ls-files` instead of requesting all files from a tree and post-filtering via expressions.
- In order to subset the list of paths (as each resolver may care about a different subset), we use wildmatch instead of regular expressions. This should be a simpler operation in most cases.

This should massively cut down the amount of file paths transferred back from gitserver, and save a lot of computation time post-filtering via regex.

## Test plan

Language-specific unit tests still pass, so there's no functional difference.

### CODE REVIEW

**Review by eseliger — COMMENTED** [review]
this looks like a good change :) thanks for following up on my report there!

File: enterprise/internal/codeintel/autoindexing/internal/inference/service.go [hunk 1]
```diff
@@ -328,24 +328,26 @@ func (s *Service) resolvePaths(
...
-	}
-
 	start := time.Now()
 	rateLimitErr := s.limiter.Wait(ctx)
 	traceLogger.AddEvent("rate_limit", attribute.Int("wait_duration_ms", int(time.Since(start).Milliseconds())))
 	if rateLimitErr != nil {
 		return nil, err
 	}
 
-	paths, err := invocationContext.gitService.ListFiles(ctx, invocationContext.repo, invocationContext.commit, pathPattern)
```
[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/service.go:343]: Actually still used in `GitTreeCodeIntelInfo`, will work on removing it there as well though.

File: enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua [hunk 2]
```diff
@@ -2,38 +2,34 @@ local patterns = require "internal_patterns"
...
 end
 
 M.new_path_literal = function(pattern)
-  return new_pattern("^", pattern, "$")
+    return new_pattern("", pattern, "")
 end
 
 M.new_path_segment = function(pattern)
-  return new_pattern("(^|/)", pattern, "(/|$)")
+    -- TODO
```
Comment (efritz) [inline enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: cc @eseliger is there a way to glob with the correct surrounding chars (or start/end of string)?

File: enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua [hunk 3]
```diff
@@ -2,38 +2,34 @@ local patterns = require "internal_patterns"
...
+    return new_pattern("*", pattern, "*")
 end
 
 M.new_path_basename = function(pattern)
-  return new_pattern("(^|/)", pattern, "$")
+    -- TODO
+    return new_pattern("*", pattern, "")
 end
```
Comment (eseliger) [inline enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:21]: what's TODO here?

Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:21]: https://github.com/sourcegraph/sourcegraph/pull/48676#discussion_r1126423028

File: enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua (hunk 2 above)
Comment (eseliger) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: For new_path_basename I'd do a simple 

```go
segments := strings.Split(path, "/")
basename := segments[len(segments)-1]
return newGlob(pattern).Match(basename)
```

for new_path_segment (this means `**/segment/**/*` right?) I'd do 

```go
segments := strings.Split(path, "/")
segmentsWithoutBasename := segments[:len(segments)-1]
for _, segment := range segments {
if segment == pathSegment {
return true
}
}
return false
```

.. and make each of the two a new return type, probably 

is that the question?

Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: Does `**/segment/**/*` work with both wildmatch and gitserver-interpreted globs? Seemed like wildmatch internals were pretty simple.

Comment (eseliger) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: no wildmatch works on a single path segment, hence the go code above to implement this in a fast manner. Wildmatch should imo only be applied to segments, and where needed. Plain string equality comparisons are still way faster.

I think git does support those, but not 100% on the details. `**/client/**` seems to return expected results in our repo for me

Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: We'll definitely want to change the code around [here](https://github.com/sourcegraph/sourcegraph/pull/48676/files#diff-e17ffd27ef75b53905e99a632e739ddf2e482b0c8feca8aa7128939fc675e085R72) then, right?

Do you think you could expose your path matching machinery from Own for re-use? Or is it not worth making it generic? If we pass a custom matcher object that could spit out one or more globs that _over_ match it, then we could still pass something to gitserver (`client/*` and `**/client/*`, for example).

Comment (eseliger) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: Probably!

Yep no problem we can share that. It's super basic, supports a pattern and then matching paths against it. It'll support `**` and `*pattern`s. Patterns should be `/` prefixed to denote it won't get a `**`. 

Is there a chance that code could stay in enterprise though? I'd love to keep that licensed if possible :D From the looks, this is all enterprise code and should be possible. 

That overmatching sounds fine, we can turn your `pattern` objects from lua into a git glob equivalent, likely

Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua:14]: Yeah all of this is enterprise.

Comment (efritz) [reply enterprise/internal/codeintel/autoindexing/internal/inference/lua/patterns.lua]: Updated!

**Review by eseliger — APPROVED** [review]

File: enterprise/internal/paths/match.go [hunk 4]
```diff
@@ -58,9 +58,13 @@ func (p asteriskPattern) Match(part string) bool {
 	return p.compiled.IsMatch(part)
 }
 
-// compile translates a text representation of a glob pattern
+type GlobPattern interface {
+	Match(filePath string) bool
+}
```
[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/internal/paths/match.go]: efcc43521fdd

File: enterprise/internal/paths/match.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package codeowners
+package paths
```
[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/internal/paths/match.go:1]: 45d587a9f05d

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/paths/match.go:1]: 🥰

**Pull request merged**